lib/core/source/common/hive_base_source.dart

```diff
@@ -0,0 +1,99 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:dartx/dartx.dart';
+import 'package:flutter_template/core/source/common/local_storage.dart';
+import 'package:hive/hive.dart';
+import 'package:rxdart/rxdart.dart';
+
+abstract class HiveBaseSource<Model, Key> implements LocalStorage<Model, Key> {
+  final Map<String, dynamic> Function(Model) dbParser;
+  final Model Function(Map<String, dynamic>) modelParser;
+  Box<String>? _box;
+
+  HiveBaseSource({required this.dbParser, required this.modelParser});
+
+  Future<Box<String>> getBox() async {
+    if (_box != null && _box!.isOpen) {
+      return _box!;
+    } else {
+      _box = await Hive.openBox<String>(Model.toString());
+      return _box!;
+    }
+  }
+
+  Future<T> withBox<T>(Future<T> Function(Box<String>) body) async =>
```

Comment:
extension?

Reply:
is it worth it? i don't want to make a new file since we will have to import it every time and moving it out of the class to a separate extension structures is kind of the same